# Data Science Pandas

## Tasks Today:

0) <b>Pre-Work</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Numpy Random Sampling

1) <b>Pandas</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Importing <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Tabular Data Structures <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - from_dict() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - read_csv() <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) <b>In-Class Exercise #1</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Accessing Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Indexing <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - df.loc <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - keys() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Slicing a DataFrame <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Built-In Methods <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - head() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - tail() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - describe() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - sort_values() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - .columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) <b>In-Class Exercise #2</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; g) Filtration <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Conditionals <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Subsetting <br>
 &nbsp;&nbsp;&nbsp;&nbsp; h) Column Transformations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Generating a New Column w/Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - User Defined Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; i) Aggregations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Type of groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - mean() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() w/Multiple Columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - drop_duplicates() <br>

## Numpy Random Sampling

In [1]:
import numpy as np

# A single call generates a single random number
print('Here is a random number: %s' % np.random.uniform())

# You can also pass some bounds
print('Here is a random number between 0 and 1 Million: %s' % np.random.uniform(0, 1e6))

# You can also generate a bunch of random numbers all at once
print('Here are 3 random numbers between 0 and 1 Million: %s' % np.random.uniform(0, 1e6, 3))

# Even matrices and tensors with weird shapes
print('Here is a 3x3 matrix of random numbers between 0 and 1 Million: \n %s' % np.random.uniform(0, 1e6, (3, 3)))

# instead of float values, generating random integers
print('Using Random Integer: %s' % np.random.randint(0, 10, 4))

Here is a random number: 0.5212757129623273
Here is a random number between 0 and 1 Million: 118419.9568454758
Here are 3 random numbers between 0 and 1 Million: [495359.02008474 666084.51924919 433124.42719005]
Here is a 3x3 matrix of random numbers between 0 and 1 Million: 
 [[788983.16268463 285300.34981486 406310.20749504]
 [790229.78508083 785883.44584812 289284.4134649 ]
 [162184.39218098 599236.87097782 861531.0872382 ]]
Using Random Integer: [5 7 3 7]


## Pandas <br>

<p>Pandas is a flexible data analysis library built on top of NumPy that is excellent for working with tabular data. It is currently the de-facto standard for Python-based data analysis, and fluency in Pandas will do wonders for your productivity and frankly your resume. It is one of the fastest ways of getting from zero to answer in existence. </p>

<ul>
    <li>Pandas is a Python module, written in C. The Pandas module is a high performance, highly efficient, and high level data analysis library. It allows us to work with large sets of data called dataframes.</li>
    <li>Series is a one-dimensional labeled array capable of holding data of any type (integer, string, float, python objects, etc.)</li>
    <li>Dataframe = Spreadsheet (has column headers, index, etc.)</li>
</ul>

### Importing

In [3]:
import pandas as pd

# always use pd, standard for data science

### Tabular data structures <br>
<p>The central object of study in Pandas is the DataFrame, which is a tabular data structure with rows and columns like an excel spreadsheet. The first point of discussion is the creation of dataframes both from native Python dictionaries, and text files through the Pandas I/O system.</p>

In [4]:
names = ['Alice',
         'Bob',
         'James',
         'Beth', 
         'John', 
         'Sally',
         'Richard', 
         'Lauren',
         'Brandon', 
         'Sabrina']

ages = np.random.randint(18,35,len(names))
my_people = {
    'names': names,
    'ages': ages
}

print(my_people)

{'names': ['Alice', 'Bob', 'James', 'Beth', 'John', 'Sally', 'Richard', 'Lauren', 'Brandon', 'Sabrina'], 'ages': array([18, 18, 21, 24, 18, 30, 32, 29, 26, 24])}


##### from_dict()

<p>Let's convert our not-so-useful-for-analysis dict into a Pandas dataframe. We can use the from_dict function to do this easily using Pandas:</p>

In [6]:
data = pd.DataFrame.from_dict(my_people)

print('Resulting type of data: {}'.format(type(data)))
data

Resulting type of data: <class 'pandas.core.frame.DataFrame'>


names  ages
0    Alice    18
1      Bob    18
2    James    21
3     Beth    24
4     John    18
5    Sally    30
6  Richard    32
7   Lauren    29
8  Brandon    26
9  Sabrina    24

##### read_csv()

In [7]:
marathon = pd.read_csv('files/boston_marathon2017_edited.csv', sep = ',')
marathon

10K                     15K                    20K  \
0      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
1      12/30/1899 12:30:27 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
2      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:16 AM   
3      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:19 AM   
4      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
...                       ...                     ...                    ...   
26405   12/30/1899 1:35:41 AM   12/30/1899 2:23:35 AM  12/30/1899 3:12:44 AM   
26406   12/30/1899 1:05:33 AM   12/30/1899 1:52:17 AM  12/30/1899 2:49:41 AM   
26407   12/30/1899 1:43:36 AM   12/30/1899 2:32:36 AM                    NaN   
26408   12/30/1899 1:27:19 AM   12/30/1899 2:17:17 AM  12/30/1899 3:11:40 AM   
26409   12/30/1899 1:17:12 AM   12/30/1899 2:00:10 AM  12/30/1899 2:58:55 AM   

                         25K                    30K                    35K  \
0      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
1      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
2      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:31 AM   
3      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:58 AM   
4      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:41 AM   
...                      ...                    ...                    ...   
26405  12/30/1899 4:12:06 AM  12/30/1899 5:03:08 AM  12/30/1899 5:55:18 AM   
26406  12/30/1899 3:50:19 AM  12/30/1899 4:50:01 AM  12/30/1899 5:53:48 AM   
26407  12/30/1899 4:15:21 AM  12/30/1899 5:06:37 AM  12/30/1899 6:00:33 AM   
26408  12/30/1899 4:06:10 AM  12/30/1899 5:07:09 AM  12/30/1899 6:06:07 AM   
26409  12/30/1899 4:27:14 AM  12/30/1899 5:37:13 AM  12/30/1899 6:39:07 AM   

                         40K                      5K  Age      Bib  ...  \
0      12/30/1899 2:02:53 AM  12/30/1899 12:15:25 AM   24     11.0  ...   
1      12/30/1899 2:03:14 AM  12/30/1899 12:15:24 AM   30     17.0  ...   
2      12/30/1899 2:03:38 AM  12/30/1899 12:15:25 AM   25     23.0  ...   
3      12/30/1899 2:04:35 AM  12/30/1899 12:15:25 AM   32     21.0  ...   
4      12/30/1899 2:05:00 AM  12/30/1899 12:15:25 AM   31      9.0  ...   
...                      ...                     ...  ...      ...  ...   
26405  12/30/1899 6:46:57 AM  12/30/1899 12:46:44 AM   61  25166.0  ...   
26406  12/30/1899 6:54:21 AM  12/30/1899 12:32:03 AM   25  25178.0  ...   
26407  12/30/1899 6:54:38 AM  12/30/1899 12:53:11 AM   57  27086.0  ...   
26408  12/30/1899 6:56:08 AM  12/30/1899 12:40:34 AM   64  25268.0  ...   
26409  12/30/1899 7:41:23 AM  12/30/1899 12:39:36 AM   48  25266.0  ...   

       Division Gender                   Half  M/F                      Name  \
0             1      1  12/30/1899 1:04:35 AM    M           Kirui, Geoffrey   
1             2      2  12/30/1899 1:04:35 AM    M               Rupp, Galen   
2             3      3  12/30/1899 1:04:36 AM    M             Osako, Suguru   
3             4      4  12/30/1899 1:04:45 AM    M          Biwott, Shadrack   
4             5      5  12/30/1899 1:04:35 AM    M            Chebet, Wilson   
...         ...    ...                    ...  ...                       ...   
26405       344  11972  12/30/1899 3:23:31 AM    F  Steinbach, Paula Eyvonne   
26406      4774  14436  12/30/1899 3:00:26 AM    M        Avelino, Andrew R.   
26407       698  11973  12/30/1899 3:36:24 AM    F           Hantel, Johanna   
26408      1043  14437  12/30/1899 3:22:30 AM    M              Reilly, Bill   
26409      2553  14438  12/30/1899 3:08:16 AM    M             Rigsby, Scott   

      Number of Records          Official Time  Overall  \
0                     1  12/30/1899 2:09:37 AM        1   
1                     1  12/30/1899 2:09:58 AM        2   
2                     1  12/30/1899 2:10:28 AM        3   
3                     1  12/30

### In-Class Exercise #1 - Read in Boston Red Sox Hitting Data <br>
<p>Use the pandas read_csv() method to read in the statistics from the two files yesterday.</p>

In [9]:
reds17 = pd.read_csv('files/redsox_2017_hitting.txt', sep = ',')
reds18 = pd.read_csv('files/redsox_2018_hitting.txt', sep = ',')
reds

Rk Pos                Name  Age    G   PA   AB    R    H  2B  ...    OBP  \
0    1   C          Sandy Leon   29   89  288  265   30   47  12  ...  0.232   
1    2  1B      Mitch Moreland   32  124  459  404   57   99  23  ...  0.325   
2    3  2B       Eduardo Nunez   31  127  502  480   56  127  23  ...  0.289   
3    4  SS     Xander Bogaerts   25  136  580  513   72  148  45  ...  0.360   
4    5  3B       Rafael Devers   21  121  490  450   59  108  24  ...  0.298   
5    6  LF   Andrew Benintendi   23  148  661  579  103  168  41  ...  0.366   
6    7  CF  Jackie Bradley Jr.   28  144  535  474   76  111  33  ...  0.314   
7    8  RF        Mookie Betts   25  136  614  520  129  180  47  ...  0.438   
8    9  DH       J.D. Martinez   30  150  649  569  111  188  37  ...  0.402   
9   10  MI          Brock Holt   30  109  367  321   41   89  18  ...  0.362   
10  11   C   Christian Vazquez   27   80  269  251   24   52  10  ...  0.257   
11  12  UT       Blake Swihart   26   82  207  192   28   44  10  ...  0.285   
12  13  UT      Hanley Ramirez   34   44  195  177   25   45   7  ...  0.313   
13  14  1B        Steve Pearce   35   50  165  136   19   38   8  ...  0.394   
14  15  2B         Ian Kinsler   36   37  143  132   17   32   6  ...  0.294   
15  16  SS         Tzu-Wei Lin   24   37   73   65   15   16   6  ...  0.329   
16  17  UT          Sam Travis   24   19   38   36    5    8   3  ...  0.263   
17  18  UT    Brandon Phillips   37    9   27   23    4    3   0  ...  0.259   
18  19  2B      Dustin Pedroia   34    3   13   11    1    1   0  ...  0.231   
19  20   C          Dan Butler   31    2    7    6    0    1   0  ...  0.143   

      SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
0   0.279  0.511    37   74    6    4   3   1    0  
1   0.433  0.758   102  175   12    0   0   5    2  
2   0.388  0.677    81  186   17    2   1   3    0  
3   0.522  0.883   135  268   14    6   0   6    4  
4   0.433  0.731    94  195    9    0   0   2    6  
5   0.465  0.830   123  269    9    2   2   7    1  
6   0.403  0.717    92  191    6   11   0   4    3  
7   0.640  1.078   186  333    5    8   0   5    8  
8   0.629  1.031   173  358   19    4   0   7   11  
9   0.411  0.774   109  132    7    7   0   2    2  
10  0.283  0.540    46   71    5    4   1   0    1  
11  0.328  0.613    65   63    4    0   0   0    0  
12  0.395  0.708    90   70    9    2   0   2    0  
13  0.507  0.901   141   69    1    5   0   2    0  
14  0.311  0.604    64   41    5    0   0   1    0  
15  0.415  0.744   100   27    0    0   0   0    0  
16  0.389  0.652    73   14    1    0   0   0    0  
17  0.261  0.520    42    6    1    0   0   0    0  
18  0.091  0.322    -7    1    0    0   0   0    0  
19  0.167  0.310   -17    1    0    0   0   1    0  

[20 rows x 28 columns]

### Accessing Data <br>

##### Indexing

<p>You can directly select a column of a dataframe just like you would a dict. The result is a Pandas 'Series' object.</p>

In [13]:
# Even though they are more complex series objects, they still support the behavior of the underlying NumPy arrays
data_ages = data['ages']
print(type(data_ages))

print(data_ages)

print(data_ages[3])

print(data['names'][0])

 # key first, row index second




<class 'pandas.core.series.Series'>
0    18
1    18
2    21
3    24
4    18
5    30
6    32
7    29
8    26
9    24
Name: ages, dtype: int64
24
Alice


##### df.loc

<p>Along the horizontal dimension, rows of Pandas DataFrames are Row objects. You will notice there is a third column present in the DataFrame - this is the $\textit{index}$. It is automatically generated as a row number, but can be reassigned to a column of your choice using the DataFrame.set_index(colname) method. We can use it to access particular Pandas $\textit{rows}$, which are also Series objects:</p>

In [16]:
print(data)
my_row = data.loc[0]
print("\n Location of first row of data: ",my_row)

print(data['names'][0])
# index after assinging the variable with .loc

# indexing without defining variable


     names  ages
0    Alice    18
1      Bob    18
2    James    21
3     Beth    24
4     John    18
5    Sally    30
6  Richard    32
7   Lauren    29
8  Brandon    26
9  Sabrina    24

 Location of first row of data:  names    Alice
ages        18
Name: 0, dtype: object


##### keys()

In [17]:
# A series can support dict-like features when it's used as a DataFrame row. 
my_row.keys()

Index(['names', 'ages'], dtype='object')

##### Slicing a DataFrame

In [19]:
 # slicing allowed
print(data[2:5])
# Slice for one item
print(data[2:3])
# The above syntax will work, but be aware, passing as single int will be interpreted as a column key
# df[2] # single indexing not allowed

   names  ages
2  James    21
3   Beth    24
4   John    18
   names  ages
2  James    21


### Built-In Methods <br>

<p>These are methods that are frequently used when using Pandas to make your life easier. It is possible to spend a whole week simply exploring the built-in functions supported by DataFrames in Pandas. Here however, we will simply highlight a few ones that might be useful, to give you an idea of what's possible out of the box with Pandas:</p>

##### head()

In [20]:
# can specify the number of rows to be displayed from the header
data.head()

names  ages
0  Alice    18
1    Bob    18
2  James    21
3   Beth    24
4   John    18

##### tail()

In [21]:
# can specify the number of rows to be displayed from the footer
data.tail()

names  ages
5    Sally    30
6  Richard    32
7   Lauren    29
8  Brandon    26
9  Sabrina    24

##### shape

In [23]:
# The dataframe has a shape property, just like a NumPy matrix. 
# print(df.shape)
print("Shape of DataFrame is: {}".format(data.shape))

# It also has an overall length property corresponding to the number of rows.
print("\n The DataFrame has a length of : {}".format(len(data)))

Shape of DataFrame is: (10, 2)

 The DataFrame has a length of : 10


##### describe() <br>
<p>Probably one of the most important methods to understand...</p>

In [26]:
# Collect summary statistics in one line
data.describe(include = 'all')

names       ages
count        10  10.000000
unique       10        NaN
top     Sabrina        NaN
freq          1        NaN
mean        NaN  24.000000
std         NaN   5.228129
min         NaN  18.000000
25%         NaN  18.750000
50%         NaN  24.000000
75%         NaN  28.250000
max         NaN  32.000000

##### sort_values()

In [28]:
# Sort based on many labels, with left-to-right priority
data.sort_values('ages', kind = 'mergesort')

names  ages
0    Alice    18
1      Bob    18
4     John    18
2    James    21
3     Beth    24
9  Sabrina    24
8  Brandon    26
7   Lauren    29
5    Sally    30
6  Richard    32

##### .columns

In [29]:
# will show all cols headers
data.columns

Index(['names', 'ages'], dtype='object')

### In-Class Exercise #2 - Describe & Sort Boston Red Sox Hitting Data <br>
<p>Take the data that you read in earlier from the Red Sox csv's and use the describe method to understand the data better. Compare the two years and decide which team is having the better year. Then sort the values based on Batting Average.</p>

In [39]:
print("boston 2017 Hitting Avg: \n {}".format(reds17['BA'].describe()['mean']))

print("boston 2018 Hitting Avg: \n {}".format(reds18['BA'].describe()['mean']))


boston 2017 Hitting Avg: 
 0.27843478260869564
boston 2018 Hitting Avg: 
 0.23795000000000002


### Filtration <br>
<p>Let's look at how to filter dataframes for rows that fulfill a specific conditon.</p>

##### Conditionals

In [41]:
# Conditional boolean dataframe
data['ages'] >= 25

0    False
1    False
2    False
3    False
4    False
5     True
6     True
7     True
8     True
9    False
Name: ages, dtype: bool

##### Subsetting

In [42]:
# exactly like numpy
data[data['ages'] >= 25]

names  ages
5    Sally    30
6  Richard    32
7   Lauren    29
8  Brandon    26

### Column Transformations <br>
<p>Rarely, if ever, will the columns in the original raw dataframe read from CSV or database table be the ones you actually need for your analysis. You will spend lots of time constantly transforming columns or groups of columns using general computational operations to produce new ones that are functions of the old ones. Pandas has full support for this: Consider the following dataframe containing membership term and renewal number for a group of customers:</p>

In [43]:
# Generate some fake data
customer_id = np.random.randint(1000,1100,10)
renewal_nbr = np.random.randint(0,10,10)
customer_dict = { 1: 0.5, 0: 1}
term_in_years = [customer_dict[key] for key in np.random.randint(0,2,10)]
# Create a dict - not usually how you would do this
random_data = {
    'customer_id':customer_id,
    'renewal_nbr':renewal_nbr,
    'term_in_years':term_in_years
}
# Build the dataframe
customers = pd.DataFrame.from_dict(random_data)
customers


customer_id  renewal_nbr  term_in_years
0         1023            1            1.0
1         1085            9            0.5
2         1096            9            0.5
3         1021            9            1.0
4         1087            0            1.0
5         1088            8            1.0
6         1064            3            0.5
7         1066            5            0.5
8         1093            4            0.5
9         1084            0            1.0

##### Generating a New Column w/Data

In [44]:
customers['customer_tenure'] = customers['renewal_nbr'] * customers['term_in_years']
customers.head(10)

customer_id  renewal_nbr  term_in_years  customer_tenure
0         1023            1            1.0              1.0
1         1085            9            0.5              4.5
2         1096            9            0.5              4.5
3         1021            9            1.0              9.0
4         1087            0            1.0              0.0
5         1088            8            1.0              8.0
6         1064            3            0.5              1.5
7         1066            5            0.5              2.5
8         1093            4            0.5              2.0
9         1084            0            1.0              0.0

##### User Defined Function

<p>If what you want to do to a column that can't be represented by simple mathematical operations, you can write your own $\textit{user defined function}$ with the full customizability available in Python and any external Python packages, then map it directly onto a column. Let's add some ages to our customer dataframe, and then classify them into our custom defined grouping scheme:</p>

In [48]:
# use .apply to map over dataframe
customers['ages'] = np.random.randint(18,70,10)

def make_age_groups(age):
    if 10 <= age and 20 > age:
        return 'Teenager'
    elif 20 <= age and 35 > age:
        return 'Young Adult'
    elif 35 <= age and 65 > age:
        return 'Adult'
    elif 65 <= age:
        return 'Senior'
    
customers['age_group'] = customers['ages']\
.apply(make_age_groups)
customers.head(10)



customer_id  renewal_nbr  term_in_years  customer_tenure  ages    age_group
0         1023            1            1.0              1.0    68       Senior
1         1085            9            0.5              4.5    59        Adult
2         1096            9            0.5              4.5    31  Young Adult
3         1021            9            1.0              9.0    33  Young Adult
4         1087            0            1.0              0.0    27  Young Adult
5         1088            8            1.0              8.0    46        Adult
6         1064            3            0.5              1.5    22  Young Adult
7         1066            5            0.5              2.5    22  Young Adult
8         1093            4            0.5              2.0    68       Senior
9         1084            0            1.0              0.0    66       Senior

<p>As a last example I'll show here how you would use a lambda function to create a UDF that depends on $\textit{more than one}$ column:</p>

<li>UDF = User Defined Function</li>

In [49]:
#Axis for apply can only be 1 or 0 -- 1 being the X axis 0 being the Y axis
def make_loyalty_age_groups(row):
    age = row['ages']
    tenure = row['customer_tenure']
    if 10 <= age and 20 > age:
        age_group = 'Teenager'
    elif 20 <= age and 35 > age:
        age_group = 'Young Adult'
    elif 35 <= age and 65 > age:
        age_group = 'Adult'
    elif 65 <= age:
        age_group = 'Senior'

    if tenure > 2.0:
        loyalty_age_group = 'loyal' + age_group
    else:
        loyalty_age_group = 'New' + age_group
        
    return loyalty_age_group

customers['loyalty_age_group'] = customers\
.apply(make_loyalty_age_groups, axis =1)

customers.head(10)

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0         1023            1            1.0              1.0    68   
1         1085            9            0.5              4.5    59   
2         1096            9            0.5              4.5    31   
3         1021            9            1.0              9.0    33   
4         1087            0            1.0              0.0    27   
5         1088            8            1.0              8.0    46   
6         1064            3            0.5              1.5    22   
7         1066            5            0.5              2.5    22   
8         1093            4            0.5              2.0    68   
9         1084            0            1.0              0.0    66   

     age_group loyalty_age_group  
0       Senior         NewSenior  
1        Adult        loyalAdult  
2  Young Adult  loyalYoung Adult  
3  Young Adult  loyalYoung Adult  
4  Young Adult    NewYoung Adult  
5        Adult        loyalAdult  
6  Young Adult    NewYoung Adult  
7  Young Adult  loyalYoung Adult  
8       Senior         NewSenior  
9       Senior         NewSenior

### In-Class Exercise #3 - Create Your Own UDF <br>
<p>Using the Boston Red Sox data, create your own UDF which creates a new column called 'All-Star' and puts every player with either a batting average over .280 or an on base percentage of over .360 with a result of 'Yes' in the column and 'No' if not.</p>

In [52]:
def check_all_star(row):
    avg = row['BA']
    obp = row['OBP']
    
    if avg > 0.28 or obp > 0.35:
        return 'Yes'
    else:
        return 'No'
    
reds18['AllStar'] = reds.apply(check_all_star, axis = 1)
reds18[['Name','BA','OBP','AllStar']]


Name     BA    OBP AllStar
0           Sandy Leon  0.177  0.232      No
1       Mitch Moreland  0.245  0.325      No
2        Eduardo Nunez  0.265  0.289      No
3      Xander Bogaerts  0.288  0.360     Yes
4        Rafael Devers  0.240  0.298      No
5    Andrew Benintendi  0.290  0.366     Yes
6   Jackie Bradley Jr.  0.234  0.314      No
7         Mookie Betts  0.346  0.438     Yes
8        J.D. Martinez  0.330  0.402     Yes
9           Brock Holt  0.277  0.362     Yes
10   Christian Vazquez  0.207  0.257      No
11       Blake Swihart  0.229  0.285      No
12      Hanley Ramirez  0.254  0.313      No
13        Steve Pearce  0.279  0.394     Yes
14         Ian Kinsler  0.242  0.294      No
15         Tzu-Wei Lin  0.246  0.329      No
16          Sam Travis  0.222  0.263      No
17    Brandon Phillips  0.130  0.259      No
18      Dustin Pedroia  0.091  0.231      No
19          Dan Butler  0.167  0.143      No

### Aggregations <br>
<p>The raw data plus some transformations is generally only half the story. Your objective is to extract actual insights and actionable conclusions from the data, and that means reducing it from potentially billions of rows to some summary statistics via aggregation functions.</p>

##### groupby() <br>
<p>The .groupby() function is in some ways a 'master' aggregation.</p> 

<p>Data tables will usually reserve one column as a primary key - that is, a column for which each row has a unique value. This is to facilitate access to the exact rows of a data table that a user wants to view. The other columns will often have repeated values, such as the age groups in the above examples. We can use these columns to explore the data using the Pandas API:</p>

In [53]:
# also introducing .count() here, exact same as to how it's used in SQL
customers.groupby('age_group', as_index = False).count().head()

age_group  customer_id  renewal_nbr  term_in_years  customer_tenure  \
0        Adult            2            2              2                2   
1       Senior            3            3              3                3   
2  Young Adult            5            5              5                5   

   ages  loyalty_age_group  
0     2                  2  
1     3                  3  
2     5                  5

##### Type of groupby()

<p>The result is a new dataframe, the columns of which all contain the counts of the grouped field. Notice the type of a grouped dataframe:</p>

In [54]:
type(customers.groupby('age_group', as_index = False))

pandas.core.groupby.generic.DataFrameGroupBy

<p>This is because simply grouping data doesn't quite make sense without an aggregation function like count() to pair with. In this case, we're counting occurances of the grouped field, but that's not all we can do. We can take averages, standard deviations, mins, maxes and much more! Let's see how this works a bit more:</p>

##### mean()

In [55]:
# mean = average
customers.groupby('age_group', as_index = False)\
.mean()[['age_group','ages']]

age_group       ages
0        Adult  52.500000
1       Senior  67.333333
2  Young Adult  27.000000

##### groupby() w/Multiple Columns

<p>We end up with the average age of the groups in the last column, the average tenure in the tenure column, and so on and so forth. You can even split the groups more finely by passing a list of columns to group by:</p>

In [56]:
customers.groupby('age_group', as_index = False)\
.mean()[['age_group','ages']]

age_group       ages
0        Adult  52.500000
1       Senior  67.333333
2  Young Adult  27.000000

##### drop_duplicates()

<p>Drops all duplicates from the current dataframe</p>

In [58]:
#cus_copy = customers

result = cus_copy.drop_duplicates()
result

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0         1023            1            1.0              1.0    68   
1         1085            9            0.5              4.5    59   
2         1096            9            0.5              4.5    31   
3         1021            9            1.0              9.0    33   
4         1087            0            1.0              0.0    27   
5         1088            8            1.0              8.0    46   
6         1064            3            0.5              1.5    22   
7         1066            5            0.5              2.5    22   
8         1093            4            0.5              2.0    68   
9         1084            0            1.0              0.0    66   

     age_group loyalty_age_group  
0       Senior         NewSenior  
1        Adult        loyalAdult  
2  Young Adult  loyalYoung Adult  
3  Young Adult  loyalYoung Adult  
4  Young Adult    NewYoung Adult  
5        Adult        loyalAdult  
6  Young Adult    NewYoung Adult  
7  Young Adult  loyalYoung Adult  
8       Senior         NewSenior  
9       Senior         NewSenior

<p>Thus the groupby operation allows you to rapidly make summary observations about the state of your entire dataset at flexible granularity. In one line above, we actually did something very complicated - that's the power of the dataframe. In fact, the process often consists of several iterative groupby operations, each revealing greater insight than the last - if you don't know where to start with a dataset, try a bunch of groupbys!</p>

### Homework Excersise #4 - Find the Mean for the Red Sox <br>
<p>For both 2017 and 2018, find the sum of HR and RBI</p>

In [91]:
# step 1: Add a new column with the key 'Team' and all column values should be 'BOS'
#groupby.sum(team)
# step 2: Group by the 'Team' column and get total home runs and rbi's

#def team_hr_rbi(row):
    #team = ['Team']
    #hr = row['HR']
    #rbi = row['RBI']
    
   # if team == team:
    #    return "BOS"
        
    
#reds18['Team'] = reds.apply(team_hr_rbi, axis = 1)
#reds18[['Team','HR','RBI',]]
#reds17.groupby('Team').sum()[['HR', 'RBI']]
reds18.groupby('Team').sum()[['HR', 'RBI']]

HR  RBI
Team          
BOS   208  826

In [81]:
 reds17.groupby(['HR', 'RBI']).sum()

Rk  Age    G   PA   AB    R    H  2B  3B  SB  ...    OBP    SLG  \
HR RBI                                                ...                 
0  0    66   82   16   17   15    3    3   1   0   0  ...  1.540  1.422   
   1    17   23   33   83   76   13   20   6   0   1  ...  0.325  0.342   
   2    57   83   63  164  150   21   40   5   2   4  ...  0.958  0.973   
   7    14   29   64  164  140   20   28   6   0   2  ...  0.305  0.243   
   9    15   28   37  118  107   10   24   2   1   1  ...  0.297  0.262   
4  12   16   30   32  108   99   10   21   2   0   0  ...  0.269  0.354   
   27   12   26   71  188  171   32   36   9   0   5  ...  0.259  0.333   
5  32    1   26   99  345  324   43   94  18   2   7  ...  0.330  0.404   
7  25   11   33   90  276  243   30   57  12   2   3  ...  0.322  0.387   
   39   10   28   85  301  271   32   61  14   0   0  ...  0.290  0.354   
   62    3   33  105  463  406   46  119  19   0   4  ...  0.369  0.392   
8  27   13   30   38  173  165   23   53  12   0   6  ...  0.353  0.539   
10 30    5   20   58  240  222   34   63  14   0   3  ...  0.338  0.482   
   62    4   24  148  635  571   94  156  32   6  15  ...  0.343  0.403   
17 63    7   27  133  541  482   58  118  19   3   8  ...  0.323  0.402   
20 90    6   22  151  658  573   84  155  26   1  20  ...  0.352  0.424   
22 79    2   31  149  576  508   73  125  34   0   0  ...  0.326  0.443   
23 62    9   33  133  553  496   58  120  24   0   1  ...  0.320  0.429   
24 102   8   24  153  712  628  101  166  46   2  26  ...  0.344  0.459   

          OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
HR RBI                                           
0  0    2.962   486    4    0    0   0   0    0  
   1    0.667    75   26    2    1   0   0    0  
   2    1.932   209   49    2    2   1   0    0  
   7    0.548    47   34    3    3   0   2    0  
   9    0.558    49   28    1    2   0   0    0  
4  12   0.622    61   35    4    0   0   1    0  
   27   0.593    54   57    8    0   3   2    0  
5  32   0.735    91  131   14    3   0   1    0  
7  25   0.709    85   94    4    2   0   1    0  
   39   0.644    68   96    5    1   1   3    1  
   62   0.760   100  159   11    2   2   4    4  
8  27   0.892   128   89    3    2   0   0    0  
10 30   0.819   111  107    5    0   0   0    3  
   62   0.746    95  230   17    6   0   2    6  
17 63   0.726    89  194    8    9   0   2    4  
20 90   0.776   102  243   16    6   1   8    7  
22 79   0.769    99  225   14    6   0   5    6  
23 62   0.750    94  213   15    6   0   0    8  
24 102  0.803   108  288    9    2   0   5    9  

[19 rows x 24 columns]